src/miniscript/context.rs

```diff
@@ -111,12 +134,31 @@ impl ScriptContext for Segwitv0 {
         frag: &Terminal<Pk, Ctx>,
     ) -> Result<(), ScriptContextError> {
         match *frag {
-            Terminal::PkK(ref pk) if pk.is_uncompressed() => {
-                Err(ScriptContextError::CompressedOnly)
+            Terminal::PkK(ref pk) => {
+                if pk.is_uncompressed() {
+                    return Err(ScriptContextError::CompressedOnly);
+                }
+
+                Ok(())
             }
             _ => Ok(()),
         }
     }
+
+    fn check_ms_validity<Pk: MiniscriptKey, Ctx: ScriptContext>(
```

Comment:
As discussed in #148, we should combine check_ms_validity and check_frag_validity

Reply:
Just implemented it and moving `check_frag_validity` to take a Miniscript reduces the number of call sites. Eg we cannot call it anymore in https://github.com/rust-bitcoin/rust-miniscript/blob/b70042e5b29b077ad123b1af5f855a02d8a4b0f9/src/miniscript/astelem.rs#L63-L75

Reply:
Actually, for all Terminal-only methods. That's the reason why i added another method in the first place: check_frag is really Terminal-only while check_ms is Miniscript-only.
At very first (#148), i combined both into check_frag and took a Terminal: it was *really* dirty.

Reply:
There is a cheap function to get Miniscript from terminals. `Miniscript::from_ast`. 
I made a commit on top of this PR that combines the checks:
https://github.com/sanket1729/rust-miniscript/commit/4311108698c3590d0f587391be5f5eee8e5e5cee

Reply:
Thanks!

Reply:
I wonder if the sanity checks in `from_ast` should be moved into `check_frag_validity`, and then `from_ast` can be replaced by an `impl From<Terminal> for Miniscript`